File: draft_exporter/html.py
from lxml import etree

from .command import Command
from .entity_state import EntityState
from .style_state import StyleState
from .wrapper_state import WrapperState


class HTML():
    def __init__(self, config):
        self.block_map = config.get('block_map', {})
        self.style_map = config.get('style_map', {})
        self.entity_decorators = config.get('entity_decorators', {})

        self.wrapper_state = WrapperState(self.block_map)

    def call(self, content_state):
        entity_map = content_state.get('entityMap', {})

        for block in content_state.get('blocks'):
            element = self.wrapper_state.element_for(block)
            self.block_contents(element, block, entity_map)

        return self.wrapper_state.to_string()

    def block_contents(self, element, block, entity_map):
        style_state = StyleState(self.style_map)
        entity_state = EntityState(element, self.entity_decorators, entity_map)
        for (text, commands) in self.build_command_groups(block):
            for command in commands:
                entity_state.apply(command)
                style_state.apply(command)

            # TODO Use entity_state.
            self.add_node(element, text, style_state)

    def add_node(self, element, text, style_state):
        if style_state.is_unstyled():
            child = etree.SubElement(element, 'textnode')
            child.text = text
        else:
            child = etree.SubElement(element, 'span', attrib=style_state.element_attributes())
            child.text = text
        # document = element.getroot()
        # node = if state.text?
        # document.create_text_node(text)
        # else
        # document.create_element('span', text, state.element_attributes)
        # end
        # element.add_child(node)

    def build_command_groups(self, block):
        """
        Creates block modification commands, grouped by start index, with the text to apply them on.
        """
        commands = self.build_commands(block)
        # TODO Tried using itertools.tee but for some reason that does not work. Oh well.
        grouped = Command.grouped_by_index(commands)
        listed = list(Command.grouped_by_index(commands))

        text = block.get('text')
        grouped_sliced = []
        i = 0
        for start_index, commands in grouped:
            next_group = listed[i + 1] if i + 1 < len(listed) else False
            stop_index = next_group[0] if next_group else 0

            grouped_sliced.append((text[start_index:stop_index], list(commands)))
            i += 1

        return grouped_sliced

    def build_commands(self, block):
        text_commands = Command.start_stop('text', 0, len(block.get('text')))
        style_commands = self.build_style_commands(block)
        entity_commands = self.build_entity_commands(block)

        return text_commands + style_commands + entity_commands

    def build_style_commands(self, block):
        return Command.from_ranges(block.get('inlineStyleRanges', []), 'inline_style', 'style')

    def build_entity_commands(self, block):
        return Command.from_ranges(block.get('entityRanges', []), 'entity', 'key')

File: draft_ruby/lib/draftjs_exporter/html.rb
# frozen_string_literal: true
require 'nokogiri'
require 'draftjs_exporter/wrapper_state'
require 'draftjs_exporter/entity_state'
require 'draftjs_exporter/style_state'
require 'draftjs_exporter/command'

module DraftjsExporter
  class HTML
    attr_reader :block_map, :style_map, :entity_decorators

    def initialize(block_map:, style_map:, entity_decorators:)
      @block_map = block_map
      @style_map = style_map
      @entity_decorators = entity_decorators
    end

    def call(content_state)
      wrapper_state = WrapperState.new(block_map)
      content_state.fetch(:blocks, []).each do |block|
        element = wrapper_state.element_for(block)
        entity_map = content_state.fetch(:entityMap, {})
        block_contents(element, block, entity_map)
      end
      wrapper_state.to_s
    end

    private

    def block_contents(element, block, entity_map)
      style_state = StyleState.new(style_map)
      entity_state = EntityState.new(element, entity_decorators, entity_map)
      build_command_groups(block).each do |text, commands|
        commands.each do |command|
          entity_state.apply(command)
          style_state.apply(command)
        end

        add_node(entity_state.current_parent, text, style_state)
      end
    end

    def add_node(element, text, state)
      document = element.document
      node = if state.text?
               document.create_text_node(text)
             else
               document.create_element('span', text, state.element_attributes)
             end
      element.add_child(node)
    end

    def build_command_groups(block)
      text = block.fetch(:text)
      grouped = build_commands(block).group_by(&:index).sort
      grouped.map.with_index { |(index, commands), command_index|
        start_index = index
        next_group = grouped[command_index + 1]
        stop_index = (next_group && next_group.first || 0) - 1
        [text.slice(start_index..stop_index), commands]
      }
    end

    def build_commands(block)
      [
        Command.new(:start_text, 0),
        Command.new(:stop_text, block.fetch(:text).size)
      ] +
        build_range_commands(:inline_style, :style, block.fetch(:inlineStyleRanges)) +
        build_range_commands(:entity, :key, block.fetch(:entityRanges))
    end

    def build_range_commands(name, data_key, ranges)
      ranges.flat_map { |range|
        data = range.fetch(data_key)
        start = range.fetch(:offset)
        stop = start + range.fetch(:length)
        [
          Command.new("start_#{name}".to_sym, start, data),
          Command.new("stop_#{name}".to_sym, stop, data)
        ]
      }
    end
  end
end
